internal/core/thirdparty/tantivy/tantivy-binding/src/analyzer/dict/lindera/fetch.rs

```diff
@@ -153,129 +155,145 @@ pub async fn fetch(
 
-    // Fast path where the data is already in cache
+    // Skip create fs locak if already in cache
     if output_dir.is_dir() {
         return Ok(());
     }
 
-    // Source file path for build package
-    let source_path_for_build = &build_dir.join(params.file_name.as_str());
-
-    // Download source file to build directory
-    let tmp_path = Path::new(&build_dir).join(params.file_name.to_owned() + ".download");
-
-    // Download a tarball
-    let client = Client::builder()
-        .user_agent(format!("Lindera/{}", env!("CARGO_PKG_VERSION")))
-        .build()?;
-
-    let mut dest = File::create(tmp_path.as_path())?;
-    let content = download_with_retry(
-        &client,
-        params.download_urls.iter().map(|s| s.as_str()).collect(),
-        MAX_ROUND,
-        params.md5_hash.as_str(),
-    )
-    .await?;
-
-    io::copy(&mut Cursor::new(content.as_slice()), &mut dest)?;
-    dest.flush()?;
-
-    rename(tmp_path.clone(), source_path_for_build)?;
-
-    // Decompress a tar.gz file
-    let tmp_extract_path = Path::new(&build_dir).join(format!("tmp-archive-{}", params.input_dir));
-    let tmp_extracted_path = tmp_extract_path.join(params.input_dir.as_str());
-    let _ = std::fs::remove_dir_all(&tmp_extract_path);
-    std::fs::create_dir_all(&tmp_extract_path)?;
-
-    let mut tar_gz = File::open(source_path_for_build)?;
-    let mut buffer = Vec::new();
-    tar_gz.read_to_end(&mut buffer)?;
-    let cursor = Cursor::new(buffer);
-    let decoder = GzDecoder::new(cursor);
-    let mut archive = Archive::new(decoder);
-    archive.unpack(&tmp_extract_path)?;
-
-    #[cfg(target_os = "windows")]
-    {
-        // Recreate input_dir to avoid conflicts when copying the directory on Windows systems (which do not support overwriting directories).
-        // Check if output_dir exists
-        if input_dir.exists() {
-            // Remove input_dir
-            std::fs::remove_dir_all(&input_dir)?;
-
-            // Make input_dir
-            std::fs::create_dir_all(&input_dir)?;
-        }
-
-        // Copy tmp_path to input_dir
-        copy_dir_all(&tmp_extracted_path, &input_dir)?;
+    let flock = File::create(&lock_path)?;
+    flock.lock_exclusive()?;
 
-        // remove tmp_path
-        std::fs::remove_dir_all(&tmp_extracted_path)?;
-    }
-    #[cfg(not(target_os = "windows"))]
-    {
-        // Empty the input directory first to avoid conflicts when renaming the directory later on Linux and macOS systems (which do not support overwriting directories).
-        empty_directory(&input_dir)?;
-        rename(tmp_extracted_path, &input_dir)?;
+    // Fast path where the data is already
+    if output_dir.is_dir() {
+        return Ok(());
     }
 
-    let _ = std::fs::remove_dir_all(&tmp_extract_path);
-    drop(dest);
-    let _ = std::fs::remove_file(source_path_for_build);
+    let build_result: Result<(), Box<dyn Error>> = {
```

Comment:
Do you plan to make this be a lambda?

Reply:
Seems unnecessary.

Reply:
> If some error with `?` returns, the following `remove_dir_all` will miss.

？just return error result to build result，remove_dir_all for input dir will still run after this block.

Reply:
Tmp file path may remain if ？returned. It won't cause any problem because will clean tmp path before use it. But remove after block expression to make sure it not be remained was better, i will update it.

Reply:
> > If some error with `?` returns, the following `remove_dir_all` will miss.
> 
> ？just return error result to build result，remove_dir_all for input dir will still run after this block.

`?` in block will return from fetch directly.